tidy3d/plugins/invdes/transformation.py

```diff
@@ -0,0 +1,73 @@
+# transformations applied to design region
+
+import typing
+import abc
+
+import jax.numpy as jnp
+import pydantic.v1 as pd
+
+import tidy3d as td
+import tidy3d.plugins.adjoint.utils.filter as tda_filter
+
+from .base import InvdesBaseModel
+
+
+class AbstractTransformation(InvdesBaseModel, abc.ABC):
+    """Base class for transformations."""
+
+
+class FilterProject(InvdesBaseModel):
+    """Transformation involving convolution by a conic filter followed by a ``tanh`` projection.
+
+    .. image:: ../../_static/img/filter_project.png
+
+    """
+
+    radius: float = pd.Field(
+        ...,
+        title="Filter Radius",
+        description="Radius of the filter to convolve with supplied spatial data. "
+        "Note: the corresponding feature size expressed in the device is typically "
+        "sqrt(3) times smaller than the radius. For best results, it is recommended to make your "
+        "radius about twice as large as the desired feature size.",
+        units=td.constants.MICROMETER,
+    )
+
+    beta: float = pd.Field(
+        1.0,
+        title="Beta",
+        description="Steepness of the binarization, "
+        "higher means more sharp transition "
+        "at the expense of gradient accuracy and ease of optimization. ",
+    )
+
+    eta: float = pd.Field(0.5, title="Eta", description="Halfway point in projection function.")
+
+    strict_binarize: bool = pd.Field(
+        False,
+        title="Binarize strictly",
+        description="If ``False``, the binarization is still continuous between min and max. "
+        "If ``True``, the values are snapped to the min and max values after projection.",
+    )
+
+    def to_filter(self, design_region_dl: float) -> tda_filter.ConicFilter:
+        """Get the conic filter associated with this transformation (after supplying grid size)."""
+        return tda_filter.ConicFilter(radius=self.radius, design_region_dl=design_region_dl)
+
+    def to_projector(self) -> tda_filter.BinaryProjector:
+        """Get the binary projector associated with this transformation."""
+        return tda_filter.BinaryProjector(
+            beta=self.beta, eta=self.eta, strict_binarize=self.strict_binarize, vmin=0, vmax=1
+        )
+
+    def evaluate(self, spatial_data: jnp.ndarray, design_region_dl: float) -> jnp.ndarray:
```

Comment:
Fine as-is, but I'm wondering if it would make sense to have the `evaluate` function output in the range [0, 1]. This would make it possible to concatenate a bunch of different parameterizations (or transformations as they are called here). So, having an arbitrary series of transformations and only converting to physical materials when they go into the simulation.

Reply:
As far as I'm aware, these `evaluate()` functions (for transformations) all output in the range (0,1) (assuming the domain is in (0,1)). are you suggesting making this explicit? ie capping the functions using `jnp.clip`?

Reply:
Uhm yeah, I think I got confused somewhere, maybe looking at the wrong file, but this is fine.